Add FNet

gchhablani opened the pull request:
# What does this PR do?

This PR adds the [FNet](https://arxiv.org/abs/2105.03824) model in PyTorch. I was working on it in another PR #12335 which got closed due to inactivity ;-;. This PR closes issue #12411.

## Checklist
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

Requesting @LysandreJik to review.

~**Note**:This model uses a SentencePiece tokenizer. They have provided the sentence-piece .model file which can be loaded. While creating FNetTokenizer should I inherit from some other existing tokenizer? Alternatively, I can copy the tokenizer from `ALBERT` (which is what I am doing right now). Wdyt?~

**Note**: I am trying to make this model as similar to Bert is possible. The original implementation has slightly different layers. For example, `FNetIntermediate` and `FNetOutput` equivalents are combined into a single layer in original FNet code, but I keep them separate. Hope this is okay?

EDIT 1:
------
I have made necessary changes for the model. And since the model compares against Bert, it makes sense to have (almost) all tasks - MultipleChoice, QuestionAnswering, etc. I am still working on:
- [x] Tokenizer (regular and fast)
- [x] Documentation
- [x] Checkpoint Conversion 
- [x] Tests

EDIT 2:
------
~We also need to skip `attention_mask` totally from the tokenizer. The user, ideally, should not have an option to get the `attention_mask` using `FNetTokenizer`. I am using `model_input_names` for this.~

EDIT 3:
------
~One more concern is that, since I am implementing in PyTorch, do we expect the user to run this on TPU? The reason is that the original implementation changes the way they calculate FFT on TPU, based on the sequence length (they found some optimal rules for faster processing). Currently, I have only used `torch.fft.fftn` directly (they use `jnp.fft.fftn` in the CPU/GPU case). Please let me know what you think.~

EDIT 4:
------
One more thing to consider is that the original code allows `type_vocab_size` of 4, which is used only for GLUE tasks. During pre-training they only use `0` and `1`. 
But, the checkpoints also have the shape of embedding weights as `(4, 768)` . Does that mean that the tokenizer might need to support something like:
```python
tokenizer = FNetTokenizer.from_pretrained('fnet-base')
inputs = tokenizer(text1, text2, text3, text4)
```
?

EDIT 5:
------
~The colab link to outputs on checkpoint conversion: [Flax to PyTorch](https://colab.research.google.com/drive/1CxxDwaH4Tei9cUBHRaMYWPHCpS2El2He?usp=sharing).
The model outputs, embedding layer, encoder layer 0 outputs match up to `1e-2`, except masked lm output for masked token, which matches to `1e-1`. Any idea on how I can improve this?~

~One reason I can think of this reduction is precision is the difference in precision in `torch.fft.fftn` and `jnp.fft.fftn` which is atmost `1e-4`. From a difference of  atmost `1e-6`, after applying the corresponding transforms, the difference becomes atmost `1e-3` in the real part. Over layers, this might accumulate. Just a guess, however.~

~This was fixed by using `gelu_new` instead of `gelu`.~

EDIT 6:
------
They use a projection layer in the embeddings, and hence the embedding size and hidden size for the model are provided separately in the config. In their experiments, they keep it same, but the flexibility is still there. Do we want to keep both the sizes separate?

EDIT 7:
------
~The FastTokenizer requires a `tokenizer.json` file which I have created using `convert_slow_tokenizer`. I used `AlbertConverter` for this model. I don't know (in-detail) how SentencePiece and FastTokenizers work. Please let me know if I'm missing anything.~

EDIT 8:
------
Just realized that the original model is denoted as `f_net`. I am using `fnet` everywhere, is this acceptable?

EDIT 9:
------
~I am not sure about special tokens in the tokenizer. The original model gives some special tokens as empty string ``. Using the current tokenizer code to load these gives `<s>` and <\s> for those tokens (bos, eos), and <**unk**> for unknown and <**pad**> for pad tokens. Not sure which is the right way to go. Any suggestions?~

gchhablani committed:
Init FNet

gchhablani committed:
Update config

gchhablani committed:
Fix config

gchhablani committed:
Update model classes

gchhablani committed:
Update tokenizers to use sentencepiece

gchhablani committed:
Fix errors in model

gchhablani committed:
Fix defaults in config

gchhablani committed:
Remove position embedding type completely

gchhablani committed:
Fix typo and take only real numbers

gchhablani committed:
Fix type vocab size in configuration

gchhablani committed:
Add projection layer to embeddings

gchhablani committed:
Fix position ids bug in embeddings

gchhablani committed:
Add minor changes

gchhablani committed:
Add conversion script and remove CausalLM vestiges

gchhablani committed:
Fix conversion script

gchhablani committed:
Fix conversion script

gchhablani committed:
Remove CausalLM Test

gchhablani committed:
Update checkpoint names to dummy checkpoints

gchhablani committed:
Add tokenizer mapping

gchhablani committed:
Fix modeling file and corresponding tests

gchhablani committed:
Add tokenization test file

gchhablani committed:
Add PreTraining model test

gchhablani committed:
Make style and quality

gchhablani committed:
Make tokenization base tests work

gchhablani committed:
Update docs

gchhablani committed:
Add FastTokenizer tests

gchhablani committed:
Fix fast tokenizer special tokens

gchhablani committed:
Fix style and quality

gchhablani committed:
Merge remote-tracking branch 'upstream/master' into add_fnet

gchhablani committed:
Remove load_tf_weights vestiges

gchhablani committed:
Add FNet to  main README

gchhablani committed:
Fix configuration example indentation

gchhablani committed:
Comment tokenization slow test

gchhablani committed:
Fix style

gchhablani commented:
Not sure why this test fails:
```python
=========================== short test summary info ============================
FAILED tests/extended/test_trainer_ext.py::TestTrainerExt::test_run_seq2seq_no_dist
==== 1 failed, 7169 passed, 3466 skipped, 708 warnings in 832.52s (0:13:52) ====

```

sgugger commented:
The failure is due to the new release of sacrebleu. If you rebase on master to get the commit that pins it to < 2.0.0, the failure will go away (but it's not necessary for this PR to be merged as we know it has nothing to do with it).

sgugger reviewed (commented):
Thanks for your PR!
I've left a few comments, but mainly the TODOs should be cleaned and all the parts that are copied from the BERT module (or another) should have Copied from statements to be checked by our internal tools.

gchhablani committed:
Merge remote-tracking branch 'upstream/master' into add_fnet

patil-suraj commented on src/transformers/models/fnet/tokenization_fnet_fast.py:
```diff
@@ -0,0 +1,203 @@
+    ) -> List[int]:
+        """
+        Build model inputs from a sequence or a pair of sequence for sequence classification tasks by concatenating and
+        adding special tokens. An ALBERT sequence has the following format:
```
```suggestion
        adding special tokens. An FNET sequence has the following format:
```

patil-suraj commented on src/transformers/models/fnet/tokenization_fnet_fast.py:
```diff
@@ -0,0 +1,203 @@
+        self, token_ids_0: List[int], token_ids_1: Optional[List[int]] = None
+    ) -> List[int]:
+        """
+        Creates a mask from the two sequences passed to be used in a sequence-pair classification task. An ALBERT
```
```suggestion
        Creates a mask from the two sequences passed to be used in a sequence-pair classification task. An FNET
```

patil-suraj reviewed (commented):
Thanks a lot for the PR, I left a few comments below.

gchhablani commented on src/transformers/models/fnet/modeling_fnet.py:
```diff
@@ -0,0 +1,1063 @@
+        return hidden_states
+
+
+class FNetFourierTransform(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        self.self = FNetBasicFourierTransform(config)
+        self.output = FNetBasicOutput(config)
+
```
@patil-suraj This is how they do it in the original code. Infact, `fft->layernorm->dense->layernorm` is done in a single module.

I wanted to keep it similar to the Bert code that's why I added it this way. But I'll change it and combine the two.

gchhablani replied:
@patil-suraj I have kept it the way it is. The `FNetBasicFourierTransform` handles the kind of fourier transform being used. `FNetFourierTransform` combines the output and transform, wdyt?

patrickvonplaten commented on docs/source/model_doc/fnet.rst:
```diff
@@ -0,0 +1,109 @@
+The FNet model was proposed in `FNet: Mixing Tokens with Fourier Transforms <https://arxiv.org/abs/2105.03824>`__ by
+James Lee-Thorp, Joshua Ainslie, Ilya Eckstein, Santiago Ontanon. The model replaces the self-attention layer in a BERT
+model with a fourier transform which returns only the real parts of the transform. The model is significantly faster
+than the BERT model because it has less parameters and is more memory efficient. The model achieves abour 92-97%
+accuracy of BERT counterparts on GLUE benchmark, and trains must faster than the BERT model. The abstract from the
+paper is the following:
```
Can we quantify "much faster" somehow? Results from the paper?

gchhablani replied:
@patrickvonplaten This is mentioned in the abstract which is added below:

> seven times faster on GPUs and twice as fast on TPUs.

patrickvonplaten commented on src/transformers/models/fnet/configuration_fnet.py:
```diff
@@ -0,0 +1,113 @@
+logger = logging.get_logger(__name__)
+
+FNET_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "gchhablani/fnet-base": "https://huggingface.co/gchhablani/fnet-base/resolve/main/config.json",
```
Let's maybe move this to the google organization no? @gchhablani - I can add you to the org, so that you can upload models under `google/...` :-)

gchhablani replied:
@patrickvonplaten Yes, that is just there temporarily for tests :p

patrickvonplaten commented on src/transformers/models/fnet/modeling_fnet.py:
```diff
@@ -0,0 +1,1063 @@
+class FNetLayer(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        # TODO: Check if chunk_size_feed_forward is needed
```
It's definitely not needed if it doesn't work out of the box - feel free to skip it

patrickvonplaten commented on src/transformers/models/fnet/modeling_fnet.py:
```diff
@@ -0,0 +1,1063 @@
+        super().__init__()
+        # TODO: Check if chunk_size_feed_forward is needed
+        self.chunk_size_feed_forward = config.chunk_size_feed_forward
+        self.seq_len_dim = 1
```
this deserves a comment IMO -> what does this mean exactly if `seq_len` has dimension 1?

gchhablani replied:
Added this:
` # The dimension which has the sequence length`

patrickvonplaten commented on src/transformers/models/fnet/modeling_fnet.py:
```diff
@@ -0,0 +1,1063 @@
+        fourier_output = self_fourier_outputs[0]
+
+        # # TODO: Check if this should be here.
+        layer_output = apply_chunking_to_forward(
```
If it works - great! Let's leave it :-) The purpose of the feed-forward chunking is explained here: https://huggingface.co/blog/reformer#2-chunked-feed-forward-layers

patrickvonplaten commented on src/transformers/models/fnet/modeling_fnet.py:
```diff
@@ -0,0 +1,1063 @@
+        )
+
+
+class FNetPooler(nn.Module):
```
A `# Copied from `BertPooler`` should also work here

patrickvonplaten commented on src/transformers/models/fnet/tokenization_fnet.py:
```diff
@@ -0,0 +1,314 @@
+        do_lower_case=True,
+        remove_space=True,
+        keep_accents=False,
+        bos_token="",
+        eos_token="",
```
You can just delete them actually

gchhablani replied:
Deleted.

patrickvonplaten commented on src/transformers/models/fnet/tokenization_fnet_fast.py:
```diff
@@ -0,0 +1,203 @@
+        do_lower_case=True,
+        remove_space=True,
+        keep_accents=False,
+        bos_token="",
```
`eos_token` and `bos_token` can be deleted here

patrickvonplaten reviewed (approved):
The PR is already in a very good state I think! 
Some things left to do:

- clean up some TODO comments
- move the weights to the google organization and add some nice model cards (I'll add you to the google org)
- Let's maybe think a bit about how we can advertise the model! Maybe just run it against `bert-base-uncased` on GLUE: https://github.com/huggingface/transformers/tree/master/examples/pytorch/text-classification#glue-tasks and then compare the training times? 

Also a team already worked with FNet during the jax/flax community sprint. Might be interesting to check their results: https://github.com/cccntu/ft5-demo

gchhablani commented on src/transformers/models/fnet/modeling_fnet.py:
```diff
@@ -0,0 +1,1063 @@
+        self.intermediate = FNetIntermediate(config)
+        self.output = FNetOutput(config)
+
+    def forward(
+        self,
+        hidden_states,
+    ):
+
+        self_fourier_outputs = self.fourier(
+            hidden_states,
+        )
```
I think it was from the template. I'm fixing all such issues.

gchhablani committed:
Add changes from review

gchhablani committed:
Merge remote-tracking branch 'upstream/master' into add_fnet

gchhablani committed:
Fix style

patrickvonplaten commented:
Hey @gchhablani :-) 

We've just added you to the Google org, so that you can move the model weights there. If you find some time, it would also be very nice to add some model cards (I can definitely help you with that). 

Regarding the failing doc test, you can just rebase to current master and it'll be fixed

gchhablani committed:
Remove bos and eos tokens from tokenizers

gchhablani commented:
Hi @patrickvonplaten @patil-suraj @sgugger @LysandreJik,

I am in a bit of a fix here. There are two issues with the fourier transform. One already existing in my code, one newly introduced (not yet pushed).

### Issue 1
The actual implementation uses `jax.vmap` on the `self.fourier_transform`. I made a mistake earlier in the implementation and do it for all dimensions - `hidden_size`, `sequence_length`, and `batch_size`, but it is just `sequence_length` and `batch_size`.

This leads to a mismatch issue when using `batch_size` more than one. I have fixed this issue by passing in the correct dimensions to `torch.fft.fftn` and using `functools.partial`.

Unfortunately, there is no `vmap` in torch as of now in the stable version, but only in the nightly version [here](https://pytorch.org/tutorials/prototype/vmap_recipe.html).

### Issue 2
Following @sgugger's suggestion to add the optimizations for TPUs, I tried adding the `einsum` version of fourier transform where they use `dft` matrix multiplication.

The problem is that `torch.einsum` does not support complex numbers and leads to a runtime error. I also tried `torch.matmul` and ran into the same issue. 

There is an issue for complex number support, and I have added a [comment](https://github.com/pytorch/pytorch/issues/33152#issuecomment-898732465)  as well. But, until then, it doesn't seem like it will be possible (unless we use jax or numpy). Please let me know how to proceed with this one. I think we can skip this, or warn somehow?

gchhablani committed:
Add tokenizer slow test, TPU transforms, NSP

gchhablani commented:
With the latest changes, an error occurs:

```python
ImportError while importing test module '/home/circleci/transformers/tests/test_modeling_fnet.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/local/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_modeling_fnet.py:23: in <module>
    from transformers.models.fnet.modeling_fnet import FNetBasicFourierTransform
src/transformers/models/fnet/modeling_fnet.py:28: in <module>
    from scipy import linalg
E   ModuleNotFoundError: No module named 'scipy'
```
I am trying to use `scipy.linalg.dft` to get `DFT matrix`. Any chance this can be a dependency?

Another error that is thrown during `test_tokenization_fnet.py`:

```python
TypeError: forward() got an unexpected keyword argument 'attention_mask'
```
This is because I removed `attention_mask` from the `tokenizer_config.json` file. How do I make it work?

gchhablani committed:
Add scipy check

gchhablani committed:
Add scipy availabilty check to test

gchhablani committed:
Fix tokenizer and use correct inputs

gchhablani committed:
Remove remaining TODOs

gchhablani committed:
Fix tests

gchhablani committed:
Fix tests

patrickvonplaten commented:
I don't see a problem with using `scipy` as an optional dependency for this specific model

patrickvonplaten commented:
Let me know if need help making the tests pass with the dependency - I can fix this in your PR if you want :-)

gchhablani commented:
Hi @patrickvonplaten

I have pushed the code where I used the global variable `_scipy_available`, does that seem okay?

Also, in model tests I'm verifying whether `fourier_transform` implementations match or not in the test: `create_and_check_fourier_transform` for which I need to access `modeling_fnet`.

I get this error on CircleCI:

```python
_________________ ERROR collecting tests/test_modeling_fnet.py _________________
ImportError while importing test module '/home/circleci/transformers/tests/test_modeling_fnet.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/local/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_modeling_fnet.py:23: in <module>
    from transformers.models.fnet.modeling_fnet import FNetBasicFourierTransform, _scipy_available
src/transformers/models/fnet/modeling_fnet.py:22: in <module>
    import torch
E   ModuleNotFoundError: No module named 'torch'
```

Any idea how do I fix this?

gchhablani committed:
Comment Fourier Test

gchhablani committed:
Uncomment Fourier Test

gchhablani committed:
Change to google checkpoint

gchhablani commented:
I was trying to use the model for `fill-mask` predictions, just so I can prepare a model card. 

The model seems to give out these outputs for the limitations and bias case:

```python
d = [
    {"sequence": "the man worked as a man.", "score": 0.07003913819789886, "token": 283, "token_str": "man"},
    {"sequence": "the man worked as a..", "score": 0.06601415574550629, "token": 16678, "token_str": "."},
    {"sequence": "the man worked as a reason.", "score": 0.020491471514105797, "token": 1612, "token_str": "reason"},
    {"sequence": "the man worked as a use.", "score": 0.017683615908026695, "token": 443, "token_str": "use"},
    {"sequence": "the man worked as a..", "score": 0.015186904929578304, "token": 845, "token_str": "."},
]

e = [
    {"sequence": "the woman worked as a..", "score": 0.12459157407283783, "token": 16678, "token_str": "."},
    {"sequence": "the woman worked as a man.", "score": 0.022601796314120293, "token": 283, "token_str": "man"},
    {"sequence": "the woman worked as a..", "score": 0.0209997296333313, "token": 845, "token_str": "."},
    {"sequence": "the woman worked as a woman.", "score": 0.01911095529794693, "token": 3806, "token_str": "woman"},
    {"sequence": "the woman worked as a one.", "score": 0.01739976927638054, "token": 276, "token_str": "one"},
]
```

And this for `A cat chases a [MASK]`:

```python
[{'sequence': 'a cat chases a cat.', 'score': 0.9358105063438416, 'token': 2882, 'token_str': 'cat'}, {'sequence': 'a cat chases a a.', 'score': 0.039954401552677155, 'token': 8, 'token_str': 'a'}, {'sequence': 'a cat chases a cats.', 'score': 0.01745556853711605, 'token': 9525, 'token_str': 'cats'}, {'sequence': 'a cat chases a..', 'score': 0.0020195094402879477, 'token': 845, 'token_str': '.'}, {'sequence': 'a cat chases a Cat.', 'score': 0.0012181063648313284, 'token': 5709, 'token_str': 'Cat'}]
```
and this for `A cat chases a [MASK]`:

```python
[{'sequence': 'a cat chases a cat.', 'score': 0.9358105063438416, 'token': 2882, 'token_str': 'cat'}, {'sequence': 'a cat chases a a.', 'score': 0.039954401552677155, 'token': 8, 'token_str': 'a'}, {'sequence': 'a cat chases a cats.', 'score': 0.01745556853711605, 'token': 9525, 'token_str': 'cats'}, {'sequence': 'a cat chases a..', 'score': 0.0020195094402879477, 'token': 845, 'token_str': '.'}, {'sequence': 'a cat chases a Cat.', 'score': 0.0012181063648313284, 'token': 5709, 'token_str': 'Cat'}]
```

:/ This doesn't feel like 92-97% accuracy of BERT.

gchhablani commented:
I have updated the checkpoints and added basic model cards. The model performance isn't great on MLM, not sure why:
- [fnet-base](https://huggingface.co/google/fnet-base)
- [fnet-large](https://huggingface.co/google/fnet-large)

LysandreJik commented on src/transformers/models/fnet/modeling_fnet.py:
```diff
@@ -0,0 +1,1186 @@
+        return seq_relationship_score
+
+
+class FNetPreTrainingHeads(nn.Module):
```
This is the same as BERT's

LysandreJik commented on src/transformers/models/fnet/modeling_fnet.py:
```diff
@@ -0,0 +1,1186 @@
+            Used to hide legacy arguments that have been deprecated.
+
+        Returns:
+
+        Example::
+            >>> from transformers import FNetTokenizer, FNetForPreTraining
```
```suggestion

        Example::
        
            >>> from transformers import FNetTokenizer, FNetForPreTraining
```

LysandreJik commented on src/transformers/models/fnet/modeling_fnet.py:
```diff
@@ -0,0 +1,1186 @@
+        effective_batch_size = input_shape[0]
+
+        #  add a dummy token
+        assert self.config.pad_token_id is not None, "The PAD token should be defined for generation"
```
We're trying to move away from assertions to `ValueError`s - do you mind replacing it here?

gchhablani replied:
This method `prepare_inputs_for_generation` is not needed, I think I forgot to remove it. Removing it now.

LysandreJik commented on src/transformers/models/fnet/modeling_fnet.py:
```diff
@@ -0,0 +1,1186 @@
+
+        Returns:
+
+        Example::
+            >>> from transformers import FNetTokenizer, FNetForNextSentencePrediction
+            >>> import torch
```
```suggestion
        Example::
        
            >>> from transformers import FNetTokenizer, FNetForNextSentencePrediction
            >>> import torch
```

LysandreJik reviewed (approved):
Thank you for your contribution @gchhablani, this looks great! I only left nitpicks.

The MLM performance may be a bit of an issue. Did you try fine-tuning it on other tasks such as sequence classification to see how it performed?

patrickvonplaten reviewed (changes requested):
Hey @gchhablani,

The model cards that you wrote
- https://huggingface.co/google/fnet-base
- https://huggingface.co/google/fnet-large

 are awesome!

It's great that you ran the fine-tuning script as well :-)

At the moment it looks like there is a bug with the MLM head though. I think we should try to fix this before merging the PR. Looking a bit closer at the MLM prediction outputs: https://huggingface.co/google/fnet-large#limitations-and-bias - something seems to be off. The mask prediction should give better results than this. Could it maybe be that the MLM weights were not correctly ported? Or it could also be a problem with the tokenizer. 

Could you maybe add some docstring that explains how the integration test can be reproduced using the original codebase like it's done for the T5 tests here: https://github.com/huggingface/transformers/blob/14e9d2954c3a7256a49a3e581ae25364c76f521e/tests/test_modeling_t5.py#L888 

This would make it much easier to verify that the `FNetForMaskedLM` weights have been correctly ported. 

Also, let me know if you'd like me to dive a bit deeper into the issue here - happy to help to get the MLM head working.

patrickvonplaten commented:
Also, just to check :-) The reported eval metrics on GLUE - did you run them once with `run_glue.py` or is it a copy-paste of the paper?

gchhablani commented:
@patrickvonplaten No, I just copy pasted from the paper 🙈. Should I try fine-tuning it?

gchhablani committed:
Add changes from review

gchhablani committed:
Merge remote-tracking branch 'upstream/master' into add_fnet

gchhablani commented:
I am checking the checkpoint conversion. Ideally, there should be less than `1e-3`/`1e-4` differences in the outputs. I'm not sure how to exactly fix this, but the arg-sorted order of the predictions is different for the PyTorch and the Flax model. :/

gchhablani committed:
Fix activation function

gchhablani committed:
Fix model integration test

gchhablani committed:
Add more integration tests

gchhablani committed:
Add comparison steps to MLM integration test

gchhablani committed:
Fix style

gchhablani commented:
The original MLM model performs poorly for this text: "the man worked as a [MASK]." The masked token top-10 predictions are:
```
s
thought
know
.
made
side
included
right
ire
now
```

However, tokenizer is having issues. The original one gives this output for the text above:
```python
[13, 283, 2479, 106, 8, 16657, 6, 16678]
[ '▁the', '▁man', '▁worked', '▁as', '▁a', '▁', '[MASK]', '.' ]
```
The tokenizer I wrote is returning this:
```python
[13, 283, 2479, 106, 8, 6, 845]
['▁the', '▁man', '▁worked', '▁as', '▁a', '[MASK]', '▁.']
```
Notice how the space - `▁` is missing and that the period is actually `.` but becomes `▁.` in our tokenizer.

Any ideas on why this might be happening?

gchhablani committed:
Add masked tokenization fix

gchhablani committed:
Improve mask tokenization fix

gchhablani commented:
I tried fixing the issue using a fix which basically skips first character after a mask token, only if it is not a `no_split_token`.

I'm not sure if this is 100% correct.

Also, there is an error with `FNetTokenizerFast`, the `[MASK]` token is not working as expected:

```python
[4, 13, 283, 2479, 106, 8, 1932, 2594, 16681, 6154, 5]
['[CLS]', '▁the', '▁man', '▁worked', '▁as', '▁a', '▁[', 'mas', 'k', '].', '[SEP]']
```

gchhablani commented:
@patrickvonplaten @LysandreJik @sgugger
Can you please check the tokenizer once when you get a chance?

patrickvonplaten commented:
Checking now!

patrickvonplaten commented on src/transformers/models/fnet/convert_fnet_original_flax_checkpoint_to_pytorch.py:
```diff
@@ -0,0 +1,155 @@
+    new_state_dict["cls.predictions.transform.LayerNorm.bias"] = torch.tensor(
+        pretrained_model_params["predictions_layer_norm"]["bias"]
+    )
+    new_state_dict["cls.predictions.decoder.weight"] = torch.tensor(
```
actually I don't think that this is needed as the output word embeddings are tied to the input word embeddings anyways

patrickvonplaten commented:
Here @gchhablani, I looked into it and the tokenizer actually looks correct to me. See this colab: https://colab.research.google.com/drive/1QC4yvSHk0DSOObD6U2fbUE-9-6W3D3_F?usp=sharing 
Note that in the original code tokens are just "manually" replaced by the "[MASK]" token. So in the colab above, if the token for "guide" (3106) is replaced by the mask token id in the original code then the current tokenizer would be correct

I'm wondering whether the model is actually the same. Do you think we could make a quick google colab check where we compare the original model with masked LM Head and the HF version? Just to make sure the outputs are 1-to-1 the same? The official google fnet weights contain all the weights for masked language modeling, so ideally we could add an integration where we make sure that `FNetForMaskedLM` works as expected

patrickvonplaten commented on tests/test_modeling_fnet.py:
```diff
@@ -0,0 +1,528 @@
+@require_torch
+class FNetModelIntegrationTest(unittest.TestCase):
+    @slow
+    def test_inference_for_masked_lm(self):
```
this is very nice!

gchhablani committed:
Merge remote-tracking branch 'upstream/master' into add_fnet

gchhablani commented:
@patrickvonplaten  The tokenizer is working as expected because of the fixed I pushed in the previous commit. It handles the mask token, but I am not 100% sure if it is correct or if there is a better way to deal with this.

gchhablani committed:
Merge branch 'master' into add_fnet

patrickvonplaten reviewed (approved):
Great work @gchhablani!

And thanks a lot for helping me on the colab. I made a colab that now shows that HF FNet and Google FNet are exactly the same: https://colab.research.google.com/drive/1hVrwWfZi0jn-U2tRRYU0UO8nxT5EOX5t?usp=sharing 

I'll send a mail to the authors to verify whether this is the intended behavior.

But looks good to merge for me now!

gchhablani commented:
@patrickvonplaten

There is an issue with `FNetTokenizerFast`:

```python
>>> from src.transformers.models.fnet.tokenization_fnet_fast import FNetTokenizerFast
2021-08-30 20:30:09.070691: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2021-08-30 20:30:09.070758: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
>>> tokenizer = FNetTokenizerFast.from_pretrained('google/fnet-base')
>>> text = "the man worked as a [MASK]."
>>> tokenizer.encode(text)
[4, 13, 283, 2479, 106, 8, 1932, 2594, 16681, 6154, 5]
>>> tokenizer.tokenize(text)
['▁the', '▁man', '▁worked', '▁as', '▁a', '▁[', 'mas', 'k', '].']
```

gchhablani committed:
Merge remote-tracking branch 'upstream/master' into add_fnet

gchhablani committed:
Fix index docs

gchhablani committed:
Merge from origin/f_net into f_net

stefan-it commented:
@gchhablani it seems that FNet was trained with a SPM vocab, so the corect masking token should be `<mask>` :(

gchhablani commented:
@stefan-it 
I haven't worked with sentencepiece before so I'm not sure. But, in the [original code](https://github.com/google-research/google-research/blob/8077479d91cca79b16417055511b7744c155c344/f_net/input_pipeline.py#L256-L258), they specify `[CLS], [SEP], [MASK]` explicitly. However, they do not use the `[MASK]` string explicitly anywhere.

What do you think about this?

If changing to `<mask>` will fix things, then we can go with it. I will try it out.

stefan-it commented:
Hi @gchhablani , oh I'm sorry I haven't yet read the official implementation. But it seems that they're really using `[MASK]` as the masking token (as previously done in [ALBERT](https://github.com/google-research/albert#sentencepiece)).

sgugger commented on docs/source/model_doc/fnet.rst:
```diff
@@ -0,0 +1,122 @@
+
+    Due to differences in the fourier transforms in PyTorch (current implementation) and JAX/Flax (original
+    implementation), the outputs of the pretrained model is not entirely identical to the original implementation. The
+    outputs for `FNetForPreTraining` match within ~0.03 of the original for MLM logits, while for NSP logits match
```
```suggestion
    outputs for :class:`~transformers.FNetForPreTraining` match within ~0.03 of the original for MLM logits, while for NSP logits match
```

sgugger commented on docs/source/model_doc/fnet.rst:
```diff
@@ -0,0 +1,122 @@
+
+.. note::
+
+    Due to differences in the fourier transforms in PyTorch (current implementation) and JAX/Flax (original
```
```suggestion
    Due to differences in the Fourier transforms in PyTorch (current implementation) and JAX/Flax (original
```

sgugger commented on src/transformers/models/fnet/modeling_fnet.py:
```diff
@@ -0,0 +1,1178 @@
+from torch.nn import CrossEntropyLoss, MSELoss
+
+
+try:
+    from scipy import linalg
+
+    _scipy_available = True
+except ImportError:
+    _scipy_available = False
```
There is an `is_scipy_available` function in file_utils that could be used.

sgugger commented on src/transformers/models/fnet/modeling_fnet.py:
```diff
@@ -0,0 +1,1178 @@
+                )
+            else:
+                logging.warning(
+                    "SciPy is need for DFT matrix calculation and is not found. Using TPU optimized fast fourier transform instead."
```
```suggestion
                    "SciPy is needed for DFT matrix calculation and is not found. Using TPU optimized fast fourier transform instead."
```

sgugger reviewed (approved):
This is in pretty good shape to be merged! Just have a few last comments.

patrickvonplaten commented:
> @patrickvonplaten
> 
> There is an issue with `FNetTokenizerFast`:
> 
> ```python
> >>> from src.transformers.models.fnet.tokenization_fnet_fast import FNetTokenizerFast
> 2021-08-30 20:30:09.070691: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
> 2021-08-30 20:30:09.070758: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
> >>> tokenizer = FNetTokenizerFast.from_pretrained('google/fnet-base')
> >>> text = "the man worked as a [MASK]."
> >>> tokenizer.encode(text)
> [4, 13, 283, 2479, 106, 8, 1932, 2594, 16681, 6154, 5]
> >>> tokenizer.tokenize(text)
> ['▁the', '▁man', '▁worked', '▁as', '▁a', '▁[', 'mas', 'k', '].']
> ```
> 
> The `[MASK]` should not get tokenized. Any idea why this might be happening?

I can fix this once the changes to how `token_type_ids` are generated are applied :-)

gchhablani committed:
Add changes from review

gchhablani committed:
Merge remote-tracking branch 'upstream/master' into add_fnet

gchhablani committed:
Fix issue

gchhablani committed:
Fix failing import in test

patrickvonplaten commented on tests/test_modeling_fnet.py:
```diff
@@ -0,0 +1,528 @@
+
+
+@require_torch
+class FNetModelIntegrationTest(unittest.TestCase):
```
Can we add one of the longer tests here that gives somewhat reasonable results (just copy paste the HF code from the colab here: https://colab.research.google.com/drive/1hVrwWfZi0jn-U2tRRYU0UO8nxT5EOX5t?usp=sharing ? Think it would be nice so that we can ensure the forward pass with seq_len = 512 stays correct :-)

patrickvonplaten replied:
just added one

gchhablani replied:
Thanks a lot @patrickvonplaten!

patrickvonplaten commented on src/transformers/models/fnet/tokenization_fnet_fast.py:
```diff
@@ -0,0 +1,199 @@
+            return cls + token_ids_0 + sep
+        return cls + token_ids_0 + sep + token_ids_1 + sep
+
+    def create_token_type_ids_from_sequences(
```
Even though the model has 4 token type ids we can leave this here! This should be fine :-)

patrickvonplaten commented on src/transformers/models/fnet/tokenization_fnet.py:
```diff
@@ -0,0 +1,441 @@
+SPIECE_UNDERLINE = "▁"
+
+
+# TODO: Need to check the special tokens once. And verify this tokenizer will work for our case.
```
Can we delete the comment here? :-)

patrickvonplaten committed:
fix merge

patrickvonplaten committed:
some more fixes

patrickvonplaten committed:
correct fast tokenizer

patrickvonplaten committed:
finalize

patrickvonplaten committed:
make style

gchhablani committed:
Merge remote-tracking branch 'upstream/master' into add_fnet

gchhablani committed:
Merge branch 'add_fnet' of https://github.com/gchhablani/transformers into add_fnet

gchhablani committed:
Remove additional tokenization logic

gchhablani committed:
Set do_lower_case to False

gchhablani committed:
Merge remote-tracking branch 'upstream/master' into add_fnet

gchhablani committed:
Allow keeping accents

gchhablani committed:
Fix tokenization test

gchhablani committed:
Fix FNet Tokenizer Fast

patrickvonplaten committed:
Merge branch 'master' of https://github.com/huggingface/transformers into add_fnet

patrickvonplaten commented on src/transformers/models/fnet/tokenization_fnet.py:
```diff
@@ -0,0 +1,304 @@
+# coding=utf-8
```
Very nice!

patrickvonplaten committed:
fix tests

patrickvonplaten committed:
make style

patrickvonplaten commented on src/transformers/pipelines/question_answering.py:
```diff
@@ -301,11 +301,14 @@ def preprocess(self, example, padding="do_not_pad", doc_stride=128, max_question
 
             features = []
             for span_idx in range(num_spans):
+                input_ids_span_idx = encoded_inputs["input_ids"][span_idx]
```
@Narsil - could you take a quick look here to check if it's ok?

gchhablani committed:
Merge remote-tracking branch 'upstream/master' into add_fnet

gchhablani committed:
Merge branch 'add_fnet' of https://github.com/gchhablani/transformers into add_fnet

gchhablani committed:
Add tips to FNet docs

gchhablani committed:
Merge remote-tracking branch 'upstream/master' into add_fnet

patrickvonplaten committed:
correct merge conflict

patrickvonplaten committed:
Merge branch 'add_fnet' of https://github.com/gchhablani/transformers into add_fnet
